src/Agent.Worker/Telemetry/TelemetryCommandExtension.cs

```diff
@@ -0,0 +1,170 @@
+using Microsoft.TeamFoundation.TestManagement.WebApi;
+using Microsoft.VisualStudio.Services.Agent.Util;
+using Microsoft.VisualStudio.Services.WebApi;
+using Microsoft.VisualStudio.Services.WebPlatform;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.Services.Agent.Worker.Telemetry
+{
+    public class TelemetryCommandExtension : AgentService, IWorkerCommandExtension
+    {
+        private long _totalThrottlingDelayInMilliseconds = 0;
+        private bool _throttlingReported = false;
+        private string _area;
+        private string _feature;
+        public HostTypes SupportedHostTypes => HostTypes.Build;
+
+        public void ProcessCommand(IExecutionContext context, Command command)
+        {
+            if (string.Equals(command.Event, WellKnownEventTrackCommand.TrackEvent, StringComparison.OrdinalIgnoreCase))
+            {
+                ProcessPublishTelemetryCommand(context, command.Properties, command.Data);
+            }
+            else
+            {
+                throw new Exception(StringUtil.Loc("TelemetryCommandNotFound", command.Event));
+            }
+        }
+
+        public Type ExtensionType
+        {
+            get
+            {
+                return typeof(IWorkerCommandExtension);
+            }
+        }
+
+        public string CommandArea
+        {
+            get
+            {
+                return "telemetry";
+            }
+        }
+
+        private void ProcessPublishTelemetryCommand(IExecutionContext context, Dictionary<string, string> eventProperties, string data)
+        {
+            ArgUtil.NotNull(context, nameof(context));
+
+            LoadTelemetryInputs(eventProperties);
+
+            var ciService = HostContext.GetService<ICustomerIntelligenceService>();
+            var vssConnection = WorkerUtilies.GetVssConnection(context, new DelegatingHandler[] { new ThrottlingReportHandler(ciService) });
+            ciService.Initialize(vssConnection);
+
+            var commandContext = HostContext.CreateService<IAsyncCommandContext>();
+            commandContext.InitializeCommandContext(context, StringUtil.Loc("Telemetry"));
+            commandContext.Task = ciService.PublishEventsAsync(new CustomerIntelligenceEvent[] { PopulateCustomerIntelligenceData(context, data) });
+            context.AsyncCommands.Add(commandContext);
+
+            // Hook up Throttling event, we will log warning on server tarpit.
+            ciService.CustomerIntelligenceQueueThrottling += ServiceThrottling_EventReceived;
+            if (_throttlingReported)
```

Comment:
will this work? you check the flag in may context, while the async command is running.